gcc/rust/util/rust-token-converter.cc

```diff
@@ -0,0 +1,586 @@
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-lex.h"
+#include "rust-token-converter.h"
+#include "libproc_macro/proc_macro.h"
+
+namespace Rust {
+
+static void
+pop_group (std::vector<ProcMacro::TokenStream> &streams,
+	   ProcMacro::Delimiter delim)
+{
+  auto g = ProcMacro::Group::make_group (streams.back (), delim);
+  streams.pop_back ();
+  auto tt = ProcMacro::TokenTree::make_tokentree (g);
+
+  streams.back ().push (tt);
+}
+
+static void
+dispatch_float_literals (ProcMacro::TokenStream &ts,
+			 const const_TokenPtr &token)
+{
+  std::string::size_type sz;
+  auto str = token->as_string ();
+  switch (token->get_type_hint ())
+    {
+      case CORETYPE_F32: {
+	auto value = std::stof (str, &sz);
+	ts.push (ProcMacro::TokenTree::make_tokentree (
+	  ProcMacro::Literal::make_f32 (value, sz == str.length ())));
+      }
+      break;
+      case CORETYPE_F64: {
+	auto value = std::stod (str, &sz);
+	ts.push (ProcMacro::TokenTree::make_tokentree (
+	  ProcMacro::Literal::make_f64 (value, sz == str.length ())));
+      }
+      break;
+    default:
+      gcc_unreachable ();
+    }
```

Comment:
@CohenArthur This comment highlight a situation that bother me a little. We're going back from an AST all the way down to a token vector. `Literals` in rust proc_macro [require me to choose a type](https://doc.rust-lang.org/proc_macro/struct.Literal.html) as there is no plain "integer"/"float" literal. Not relying on a the `CORETYPE` hint means I would have to set a "default" (i32 and f64) type which could then be promoted to a different "true" type. While this could potentially work I'm not entirely convinced by this approach, something feels a bit off.

Reply:
Looks likeI was wrong on this one, the proc_macro interface does not expose it but there is an `Integer` type.

For the following input:
```rust
pub fn default_type() { let a = 3 ; }
```
We get the following dump:
```
(type: TokenStream [Ident { ident: "pub", span: #0 bytes(387..390) }, Ident { ident: "fn", span: #0 bytes(391..393) }, Ident { ident: "default_type", span: #0 bytes(394..406) }, Group { delimiter: Parenthesis, stream: TokenStream [], span: #0 bytes(406..408) }, Group { delimiter: Brace, stream: TokenStream [Ident { ident: "let", span: #0 bytes(415..418) }, Ident { ident: "a", span: #0 bytes(419..420) }, Punct { ch: '=', spacing: Alone, span: #0 bytes(421..422) }, Literal { kind: Integer, symbol: "3", suffix: None, span: #0 bytes(423..424) }, Punct { ch: ';', spacing: Alone, span: #0 bytes(424..425) }], span: #0 bytes(409..427) }])
```

where we can find the following:

```
Literal { kind: Integer, symbol: "3", suffix: None, span: #0 bytes(423..424) }
```